File: sdk-java/src/test/java/io/superdurable/dex/integ/WorkflowUncompletedTest.java
/*
 * Portions of this file are derived from indeedeng/iwf-java-sdk.
 * Those portions are licensed under the Apache License, Version 2.0.
 * See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
 *
 * Modifications Copyright (c) 2026 Super Durable, Inc.
 *
 * Modifications are licensed under the Super Durable Source License 1.0.
 * Third-Party Materials remain under the Apache License, Version 2.0.
 * See LICENSE and LEGACY_NOTICES.md.
 */

package io.superdurable.dex.integ;

import io.superdurable.dex.Client;
import io.superdurable.dex.Context;
import io.superdurable.dex.Flow;
import io.superdurable.dex.FlowErrorType;
import io.superdurable.dex.FlowResult;
import io.superdurable.dex.FlowStatus;
import io.superdurable.dex.FlowTimeoutPolicy;
import io.superdurable.dex.exceptions.LongPollTimeoutException;
import io.superdurable.dex.StartFlowOptions;
import io.superdurable.dex.StepDecision;
import io.superdurable.dex.StopFlowOptions;
import io.superdurable.dex.StopType;
import io.superdurable.dex.testing.DexDevTestEnvironment;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import java.nio.file.Path;
import java.time.Duration;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

@Tag("dex-dev")
public final class WorkflowUncompletedTest {
    private static final SignalWorkflow WAIT_TIMEOUT_WORKFLOW = new SignalWorkflow();
    private static final WorkflowUncompletedForceFailWorkflow FORCE_FAIL_WORKFLOW =
            new WorkflowUncompletedForceFailWorkflow();
    private static final WorkflowUncompletedStateFailureWorkflow STATE_FAILURE_WORKFLOW =
            new WorkflowUncompletedStateFailureWorkflow();
    private static final WorkflowUncompletedStateTimeoutWorkflow STATE_TIMEOUT_WORKFLOW =
            new WorkflowUncompletedStateTimeoutWorkflow();
    private static final WorkflowUncompletedEmptyDecisionWorkflow EMPTY_DECISION_WORKFLOW =
            new WorkflowUncompletedEmptyDecisionWorkflow();

    @TempDir
    Path cacheDirectory;

    @Test
    void testFlowWaitTimeout() throws Exception {
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                WAIT_TIMEOUT_WORKFLOW)) {
            final String flowId = flowId("wait-timeout");
            environment.client().startFlow(WAIT_TIMEOUT_WORKFLOW, flowId, 1);

            final LongPollTimeoutException failure = assertThrows(
                    LongPollTimeoutException.class,
                    () -> environment.client().waitForFlow(flowId, Duration.ofSeconds(1)).getSingleOutput(Integer.class));
            assertEquals(flowId, failure.getFlowId());
        }
    }

    @Test
    void testFlowTimeout() throws Exception {
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                WAIT_TIMEOUT_WORKFLOW)) {
            final String flowId = flowId("flow-timeout");
            environment.client().startFlow(
                    WAIT_TIMEOUT_WORKFLOW,
                    flowId,
                    1,
                    StartFlowOptions.newBuilder().timeout(Duration.ofSeconds(1)).build());

            final FlowResult failure = waitForFailure(environment, flowId);
            assertFailure(
                    failure,
                    FlowStatus.FAILED,
                    FlowErrorType.FLOW_TIMEOUT,
                    "Flow timed out after 1 seconds",
                    0);
        }
    }

    @Test
    void testFlowTimeoutHandler() throws Exception {
        final TimeoutHandlerFlow flow = new TimeoutHandlerFlow();
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                flow)) {
            final String flowId = flowId("flow-timeout-handler");
            environment.client().startFlow(
                    flow,
                    flowId,
                    null,
                    StartFlowOptions.newBuilder().timeout(Duration.ofSeconds(1)).build());

            assertEquals(
                    "expired",
                    environment.client()
                            .waitForFlow(flowId, Duration.ofSeconds(15))
                            .getSingleOutput(String.class));
        }
    }

    @Test
    void testFlowTimeoutHandlerCancelOverride() throws Exception {
        final TimeoutHandlerFlow flow = new TimeoutHandlerFlow();
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                flow)) {
            final String flowId = flowId("flow-timeout-handler-cancel");
            environment.client().startFlow(
                    flow,
                    flowId,
                    null,
                    StartFlowOptions.newBuilder()
                            .timeout(Duration.ofSeconds(1))
                            .timeoutPolicy(FlowTimeoutPolicy.CANCEL)
                            .build());

            assertFailure(
                    waitForFailure(environment, flowId),
                    FlowStatus.CANCELED,
                    null,
                    null,
                    0);
        }
    }

    @Test
    void testFlowTimeoutHandlerRequiresOverride() throws Exception {
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                WAIT_TIMEOUT_WORKFLOW)) {
            final IllegalArgumentException missingTimeout = assertThrows(
                    IllegalArgumentException.class,
                    () -> environment.client().startFlow(
                            WAIT_TIMEOUT_WORKFLOW,
                            flowId("flow-timeout-policy-without-timeout"),
                            1,
                            StartFlowOptions.newBuilder()
                                    .timeoutPolicy(FlowTimeoutPolicy.CANCEL)
                                    .build()));
            assertTrue(missingTimeout.getMessage().contains("requires a positive timeout"));

            final IllegalArgumentException failure = assertThrows(
                    IllegalArgumentException.class,
                    () -> environment.client().startFlow(
                            WAIT_TIMEOUT_WORKFLOW,
                            flowId("flow-timeout-handler-missing"),
                            1,
                            StartFlowOptions.newBuilder()
                                    .timeout(Duration.ofSeconds(1))
                                    .timeoutPolicy(FlowTimeoutPolicy.HANDLER)
                                    .build()));
            assertTrue(failure.getMessage().contains("does not override handleTimeout"));
        }
    }

    @Test
    void testFlowCanceled() throws Exception {
        assertStoppedFlow(StopType.CANCEL, null, FlowStatus.CANCELED, null, null);
    }

    @Test
    void testFlowTerminated() throws Exception {
        assertStoppedFlow(
                StopType.TERMINATE,
                "terminated",
                FlowStatus.TERMINATED,
                null,
                null);
    }

    @Test
    void testFlowFailedByApi() throws Exception {
        assertStoppedFlow(
                StopType.FAIL,
                "fail by API",
                FlowStatus.FAILED,
                FlowErrorType.CLIENT_API_FAILED,
                "fail by API");
    }

    @Test
    void testForceFailFlow() throws Exception {
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                FORCE_FAIL_WORKFLOW)) {
            final String flowId = flowId("force-fail");
            environment.client().startFlow(
                    FORCE_FAIL_WORKFLOW,
                    flowId,
                    5);

            final FlowResult failure = waitForFailure(environment, flowId);
            assertFailure(
                    failure,
                    FlowStatus.FAILED,
                    FlowErrorType.STEP_DECISION_FAILED,
                    "a failing message",
                    0);
        }
    }

    @Test
    void testWorkerApiFailure() throws Exception {
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                STATE_FAILURE_WORKFLOW)) {
            final String flowId = flowId("worker-api-failure");
            environment.client().startFlow(
                    STATE_FAILURE_WORKFLOW,
                    flowId,
                    5);

            final FlowResult failure = waitForFailure(environment, flowId);
            assertEquals(FlowStatus.FAILED, failure.getStatus());
            assertEquals(FlowErrorType.WORKER_API_FAILED, failure.getErrorType());
            assertTrue(failure.getErrorMessage().contains("test api failing"), failure.getErrorMessage());
            assertEquals(0, failure.getCompletions().size());
        }
    }

    @Test
    void testWorkerApiTimeout() throws Exception {
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                STATE_TIMEOUT_WORKFLOW)) {
            final String flowId = flowId("worker-api-timeout");
            environment.client().startFlow(
                    STATE_TIMEOUT_WORKFLOW,
                    flowId,
                    5);

            final FlowResult failure = waitForFailure(environment, flowId);
            assertEquals(FlowStatus.FAILED, failure.getStatus());
            assertEquals(FlowErrorType.WORKER_API_FAILED, failure.getErrorType());
            assertTrue(failure.getErrorMessage().toLowerCase().contains("timeout"), failure.getErrorMessage());
            assertEquals(0, failure.getCompletions().size());
        }
    }

    @Test
    void testEmptyDecisionFailsFlow() throws Exception {
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                EMPTY_DECISION_WORKFLOW)) {
            final String flowId = flowId("empty-decision");
            environment.client().startFlow(
                    EMPTY_DECISION_WORKFLOW,
                    flowId,
                    5);

            final FlowResult failure = waitForFailure(environment, flowId);
            assertEquals(FlowStatus.FAILED, failure.getStatus());
            assertEquals(FlowErrorType.WORKER_API_FAILED, failure.getErrorType());
            assertTrue(
                    failure.getErrorMessage().contains("goToMulti requires a movement"),
                    failure.getErrorMessage());
            assertEquals(0, failure.getCompletions().size());
        }
    }

    private void assertStoppedFlow(
            final StopType stopType,
            final String reason,
            final FlowStatus expectedStatus,
            final FlowErrorType expectedErrorType,
            final String expectedMessage) throws Exception {
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                WAIT_TIMEOUT_WORKFLOW)) {
            final String flowId = flowId("stopped");
            environment.client().startFlow(
                    WAIT_TIMEOUT_WORKFLOW,
                    flowId,
                    1);
            environment.client().stopFlow(flowId, new StopFlowOptions(stopType, reason));

            final FlowResult failure = waitForFailure(environment, flowId);
            assertFailure(
                    failure,
                    expectedStatus,
                    expectedErrorType,
                    expectedMessage,
                    0);
        }
    }

    private static FlowResult waitForFailure(
            final DexDevTestEnvironment environment,
            final String flowId) {
        return environment.client().waitForFlow(flowId, Duration.ofSeconds(15));
    }

    private static void assertFailure(
            final FlowResult failure,
            final FlowStatus status,
            final FlowErrorType errorType,
            final String message,
            final int resultCount) {
        assertEquals(status, failure.getStatus());
        assertEquals(errorType, failure.getErrorType());
        if (message == null) {
            assertNull(failure.getErrorMessage());
        } else {
            assertEquals(message, failure.getErrorMessage());
        }
        assertEquals(resultCount, failure.getCompletions().size());
    }

    private static String flowId(final String prefix) {
        return prefix + "-" + UUID.randomUUID();
    }

    private static final class TimeoutHandlerFlow implements Flow<Void> {
        @Override
        public StepDecision handleTimeout(final Context context) {
            return StepDecision.forceComplete("expired");
        }
    }
}

File: sdk-rust/crates/dex-sdk/tests/integ/workflow_uncompleted_test.rs
// Portions of this file are derived from indeedeng/iwf-java-sdk.
// Those portions are licensed under the Apache License, Version 2.0.
// See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
//
// Modifications Copyright (c) 2026 Super Durable, Inc.
//
// Modifications are licensed under the Super Durable Source License 1.0.
// Third-Party Materials remain under the Apache License, Version 2.0.
// See LICENSE and LEGACY_NOTICES.md.

use std::sync::LazyLock;
use std::time::Duration;

use dex_sdk::{
    Client, Context, Flow, FlowErrorType, FlowStatus, FlowTimeoutHandler, FlowTimeoutPolicy,
    HandlerResult, PersistenceSchema, Registry, SdkError, SdkResult, StartFlowOptions,
    StepDecision, StopFlowOptions, Stream,
};

use crate::signal_workflow::SignalWorkflow;
use crate::support::{DexDevTestEnvironment, flow_id};
use crate::workflow_uncompleted_empty_decision_workflow::WorkflowUncompletedEmptyDecisionWorkflow;
use crate::workflow_uncompleted_force_fail_workflow::WorkflowUncompletedForceFailWorkflow;
use crate::workflow_uncompleted_state_failure_workflow::WorkflowUncompletedStateFailureWorkflow;
use crate::workflow_uncompleted_state_timeout_workflow::WorkflowUncompletedStateTimeoutWorkflow;

#[test]
#[ignore = "requires dexcli dev"]
fn test_flow_wait_timeout() {
    let environment = DexDevTestEnvironment::start(Registry::new().register(SignalWorkflow::new()));
    let workflow = SignalWorkflow::new();
    let flow_id = flow_id("wait-timeout");
    environment
        .client
        .start_flow(&workflow, &flow_id, 1)
        .expect("start waiting Flow");
    match environment
        .client
        .wait_for_flow_with_timeout(&flow_id, Duration::from_secs(1))
        .and_then(|result| result.single_output::<i32>())
        .expect_err("waitForFlow must time out")
    {
        SdkError::LongPollTimeout { service } => {
            assert_eq!(Some(flow_id.as_str()), service.flow_id())
        }
        error => panic!("expected LongPollTimeout, got {error:?}"),
    }
}

#[test]
#[ignore = "requires dexcli dev"]
fn test_flow_timeout() {
    let environment = DexDevTestEnvironment::start(Registry::new().register(SignalWorkflow::new()));
    let workflow = SignalWorkflow::new();
    let flow_id = flow_id("flow-timeout");
    environment
        .client
        .start_flow_with_options(
            &workflow,
            &flow_id,
            1,
            StartFlowOptions::new().timeout(Duration::from_secs(1)),
        )
        .expect("start timed Flow");
    assert_failure(
        wait_for_failure(&environment, &flow_id),
        FlowStatus::Failed,
        Some(FlowErrorType::FlowTimeout),
        Some("Flow timed out after 1 seconds"),
        0,
    );
}

struct TimeoutHandlerFlow;

static TIMEOUT_PROGRESS: LazyLock<Stream<String>> =
    LazyLock::new(|| Stream::new("timeout-progress", 1 << 20));

impl TimeoutHandlerFlow {
    fn handle_timeout(&self, context: &mut Context) -> HandlerResult<StepDecision> {
        assert!(context.record_heartbeat().is_err());
        assert!(
            TIMEOUT_PROGRESS
                .write(context, "invalid".to_string())
                .is_err()
        );
        Ok(StepDecision::force_complete("expired".to_string()))
    }
}

impl Flow for TimeoutHandlerFlow {
    type StartInput = ();

    fn persistence(&self) -> PersistenceSchema {
        PersistenceSchema::new().stream(&TIMEOUT_PROGRESS)
    }

    fn timeout_handler(&self) -> Option<FlowTimeoutHandler<Self>> {
        Some(Self::handle_timeout)
    }
}

#[test]
#[ignore = "requires dexcli dev"]
fn test_flow_timeout_handler() {
    let environment = DexDevTestEnvironment::start(Registry::new().register(TimeoutHandlerFlow));
    let flow_id = flow_id("flow-timeout-handler");
    environment
        .client
        .start_flow_with_options(
            &TimeoutHandlerFlow,
            &flow_id,
            (),
            StartFlowOptions::new().timeout(Duration::from_secs(1)),
        )
        .expect("start timeout-handler Flow");
    let output: String = environment
        .client
        .wait_for_flow_with_timeout(&flow_id, Duration::from_secs(15))
        .and_then(|result| result.single_output())
        .expect("timeout handler completes Flow");
    assert_eq!("expired", output);
}

#[test]
#[ignore = "requires dexcli dev"]
fn test_flow_timeout_handler_cancel_override() {
    let environment = DexDevTestEnvironment::start(Registry::new().register(TimeoutHandlerFlow));
    let flow_id = flow_id("flow-timeout-handler-cancel");
    environment
        .client
        .start_flow_with_options(
            &TimeoutHandlerFlow,
            &flow_id,
            (),
            StartFlowOptions::new()
                .timeout(Duration::from_secs(1))
                .timeout_policy(FlowTimeoutPolicy::Cancel),
        )
        .expect("start canceled timeout Flow");
    assert_failure(
        wait_for_failure(&environment, &flow_id),
        FlowStatus::Canceled,
        None,
        None,
        0,
    );
}

#[test]
#[ignore = "requires dexcli dev"]
fn test_flow_timeout_handler_requires_registration() {
    let environment = DexDevTestEnvironment::start(Registry::new().register(SignalWorkflow::new()));
    let workflow = SignalWorkflow::new();
    let error = environment
        .client
        .start_flow_with_options(
            &workflow,
            &flow_id("flow-timeout-policy-without-timeout"),
            1,
            StartFlowOptions::new().timeout_policy(FlowTimeoutPolicy::Cancel),
        )
        .expect_err("non-default policy must require a positive timeout");
    assert!(matches!(
        error,
        SdkError::InvalidArgument { message } if message.contains("requires a positive timeout")
    ));

    let error = environment
        .client
        .start_flow_with_options(
            &workflow,
            &flow_id("flow-timeout-handler-missing"),
            1,
            StartFlowOptions::new()
                .timeout(Duration::from_secs(1))
                .timeout_policy(FlowTimeoutPolicy::Handler),
        )
        .expect_err("handler policy must require a registered handler");
    assert!(matches!(
        error,
        SdkError::InvalidArgument { message } if message.contains("has no timeout handler")
    ));
}

#[test]
#[ignore = "requires dexcli dev"]
fn test_flow_canceled() {
    assert_stopped_flow(StopFlowOptions::cancel(), FlowStatus::Canceled, None, None);
}

#[test]
#[ignore = "requires dexcli dev"]
fn test_flow_terminated() {
    assert_stopped_flow(
        StopFlowOptions::terminate().reason("terminated"),
        FlowStatus::Terminated,
        None,
        None,
    );
}

#[test]
#[ignore = "requires dexcli dev"]
fn test_flow_failed_by_api() {
    assert_stopped_flow(
        StopFlowOptions::fail().reason("fail by API"),
        FlowStatus::Failed,
        Some(FlowErrorType::ClientApiFailed),
        Some("fail by API"),
    );
}

#[test]
#[ignore = "requires dexcli dev"]
fn test_force_fail_flow() {
    let environment = DexDevTestEnvironment::start(
        Registry::new().register(WorkflowUncompletedForceFailWorkflow::new()),
    );
    let workflow = WorkflowUncompletedForceFailWorkflow::new();
    let flow_id = flow_id("force-fail");
    environment
        .client
        .start_flow(&workflow, &flow_id, 5)
        .expect("start force-fail Flow");
    assert_failure(
        wait_for_failure(&environment, &flow_id),
        FlowStatus::Failed,
        Some(FlowErrorType::StepDecisionFailed),
        Some("a failing message"),
        0,
    );
}

#[test]
#[ignore = "requires dexcli dev"]
fn test_worker_api_failure() {
    let environment = DexDevTestEnvironment::start(
        Registry::new().register(WorkflowUncompletedStateFailureWorkflow::new()),
    );
    let workflow = WorkflowUncompletedStateFailureWorkflow::new();
    let flow_id = flow_id("worker-api-failure");
    environment
        .client
        .start_flow(&workflow, &flow_id, 5)
        .expect("start worker-failure Flow");
    let result = wait_for_failure(&environment, &flow_id);
    assert_eq!(FlowStatus::Failed, result.status());
    assert_eq!(Some(FlowErrorType::WorkerApiFailed), result.error_type());
    assert!(
        result
            .error_message()
            .is_some_and(|value| value.contains("test api failing"))
    );
    assert_eq!(0, result.completions().len());
}

#[test]
#[ignore = "requires dexcli dev"]
fn test_worker_api_timeout() {
    let environment = DexDevTestEnvironment::start(
        Registry::new().register(WorkflowUncompletedStateTimeoutWorkflow::new()),
    );
    let workflow = WorkflowUncompletedStateTimeoutWorkflow::new();
    let flow_id = flow_id("worker-api-timeout");
    environment
        .client
        .start_flow(&workflow, &flow_id, 5)
        .expect("start worker-timeout Flow");
    let result = wait_for_failure(&environment, &flow_id);
    assert_eq!(FlowStatus::Failed, result.status());
    assert_eq!(Some(FlowErrorType::WorkerApiFailed), result.error_type());
    assert!(
        result
            .error_message()
            .is_some_and(|value| value.to_lowercase().contains("timeout"))
    );
    assert_eq!(0, result.completions().len());
}

#[test]
#[ignore = "requires dexcli dev"]
fn test_empty_decision_fails_flow() {
    let environment = DexDevTestEnvironment::start(
        Registry::new().register(WorkflowUncompletedEmptyDecisionWorkflow::new()),
    );
    let workflow = WorkflowUncompletedEmptyDecisionWorkflow::new();
    let flow_id = flow_id("empty-decision");
    environment
        .client
        .start_flow(&workflow, &flow_id, 5)
        .expect("start empty-decision Flow");
    let result = wait_for_failure(&environment, &flow_id);
    assert_eq!(FlowStatus::Failed, result.status());
    assert_eq!(Some(FlowErrorType::WorkerApiFailed), result.error_type());
    assert!(
        result
            .error_message()
            .is_some_and(|value| value.contains("go_to_many requires a movement"))
    );
    assert_eq!(0, result.completions().len());
}

fn assert_stopped_flow(
    options: StopFlowOptions,
    expected_status: FlowStatus,
    expected_error_type: Option<FlowErrorType>,
    expected_message: Option<&str>,
) {
    let environment = DexDevTestEnvironment::start(Registry::new().register(SignalWorkflow::new()));
    let workflow = SignalWorkflow::new();
    let flow_id = flow_id("stopped");
    environment
        .client
        .start_flow(&workflow, &flow_id, 1)
        .expect("start stoppable Flow");
    environment
        .client
        .stop_flow(&flow_id, options)
        .expect("stop Flow");
    assert_failure(
        wait_for_failure(&environment, &flow_id),
        expected_status,
        expected_error_type,
        expected_message,
        0,
    );
}

fn wait_for_failure(environment: &DexDevTestEnvironment, flow_id: &str) -> dex_sdk::FlowResult {
    environment
        .client
        .wait_for_flow_with_timeout(flow_id, Duration::from_secs(15))
        .expect("wait for terminal Flow result")
}

fn assert_failure(
    failure: dex_sdk::FlowResult,
    expected_status: FlowStatus,
    expected_error_type: Option<FlowErrorType>,
    expected_message: Option<&str>,
    expected_result_count: usize,
) {
    assert_eq!(expected_status, failure.status());
    assert_eq!(expected_error_type, failure.error_type());
    assert_eq!(expected_message, failure.error_message());
    assert_eq!(expected_result_count, failure.completions().len());
}

#[allow(dead_code)]
fn compile_workflow_uncompleted(client: &Client) -> SdkResult<()> {
    client.start_flow(
        &WorkflowUncompletedForceFailWorkflow::new(),
        "force-fail",
        5,
    )?;
    client.start_flow(
        &WorkflowUncompletedStateFailureWorkflow::new(),
        "worker-failure",
        5,
    )?;
    client.start_flow(
        &WorkflowUncompletedStateTimeoutWorkflow::new(),
        "worker-timeout",
        5,
    )?;
    client.start_flow(
        &WorkflowUncompletedEmptyDecisionWorkflow::new(),
        "empty-decision",
        5,
    )?;
    Ok(())
}
